File: proj.android_studio/ee-x-firebase-performance/src/main/java/com/ee/internal/FirebasePerformanceTrace.kt
package com.ee.internal

import androidx.annotation.AnyThread
import com.ee.IMessageBridge
import com.google.firebase.perf.metrics.Trace
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.Serializable

@InternalSerializationApi
internal class FirebasePerformanceTrace(
    private val _bridge: IMessageBridge,
    private val _trace: Trace,
    private val _traceName: String) {
    companion object {
        private const val kPrefix = "FirebasePerformance"
    }

    init {
        registerHandlers()
    }

    fun destroy() {
        deregisterHandlers()
    }

    private val kStart: String
        @AnyThread get() = "${kPrefix}_start_$_traceName"

    private val kStop: String
        @AnyThread get() = "${kPrefix}_stop_$_traceName"

    private val kIncrementMetric: String
        @AnyThread get() = "${kPrefix}_incrementMetric_$_traceName"

    private val kGetLongMetric: String
        @AnyThread get() = "${kPrefix}_getLongMetric_$_traceName"

    private val kPutMetric: String
        @AnyThread get() = "${kPrefix}_putMetric_$_traceName"

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(kStart) {
            start()
            ""
        }
        _bridge.registerHandler(kStop) {
            stop()
            ""
        }
        _bridge.registerHandler(kIncrementMetric) { message ->
            @Serializable
            class Request(
                val key: String,
                val value: Long
            )

            val request = deserialize<Request>(message)
            incrementMetric(request.key, request.value)
            ""
        }
        _bridge.registerHandler(kPutMetric) { message ->
            @Serializable
            class Request(
                val key: String,
                val value: Long
            )

            val request = deserialize<Request>(message)
            putMetric(request.key, request.value)
            ""
        }
        _bridge.registerHandler(kGetLongMetric) { message ->
            getLongMetric(message).toString()
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kStart)
        _bridge.deregisterHandler(kStop)
        _bridge.deregisterHandler(kIncrementMetric)
        _bridge.deregisterHandler(kPutMetric)
        _bridge.deregisterHandler(kGetLongMetric)
    }

    @AnyThread
    fun start() {
        _trace.start()
    }

    @AnyThread
    fun stop() {
        _trace.stop()
    }

    @AnyThread
    private fun incrementMetric(traceName: String, value: Long) {
        _trace.incrementMetric(traceName, value)
    }

    @AnyThread
    private fun putMetric(name: String, value: Long) {
        _trace.putMetric(name, value)
    }

    @AnyThread
    private fun getLongMetric(name: String): Long {
        return _trace.getLongMetric(name)
    }
}

File: src/ios/ee/firebase_performance/FirebasePerformanceTrace.swift
//
//  FirebasePerformanceTrace.swift
//  ee-x-bb8dc50f
//
//  Created by eps on 2/3/21.
//

private let kPrefix = "FirebasePerformance"

internal class FirebasePerformanceTrace {
    private let _bridge: IMessageBridge
    private let _traceName: String
    private let _trace: Trace

    public init(_ bridge: IMessageBridge, _ traceName: String, _ trace: Trace) {
        _bridge = bridge
        _traceName = traceName
        _trace = trace
        registerHandlers()
    }

    public func destroy() {
        deregisterHandlers()
    }

    private var kStart: String {
        return "\(kPrefix)_start_\(_traceName)"
    }

    private var kStop: String {
        return "\(kPrefix)_stop_\(_traceName)"
    }

    private var kIncrementMetric: String {
        return "\(kPrefix)_incrementMetric_\(_traceName)"
    }

    private var kGetLongMetric: String {
        return "\(kPrefix)_getLongMetric_\(_traceName)"
    }

    private var kPutMetric: String {
        return "\(kPrefix)_putMetric_\(_traceName)"
    }

    func registerHandlers() {
        _bridge.registerHandler(kStart) { _ in
            self.start()
            return ""
        }
        _bridge.registerHandler(kStop) { _ in
            self.stop()
            return ""
        }
        _bridge.registerHandler(kIncrementMetric) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard
                let key = dict["key"] as? String,
                let value = dict["value"] as? Int64
            else {
                assert(false, "Invalid argument")
                return ""
            }
            self.incrementMetric(key, value)
            return ""
        }
        _bridge.registerHandler(kPutMetric) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard
                let key = dict["key"] as? String,
                let value = dict["value"] as? Int64
            else {
                assert(false, "Invalid argument")
                return ""
            }
            self.putMetric(key, value)
            return ""
        }
        _bridge.registerHandler(kGetLongMetric) { message in
            "\(self.getLongMetric(message))"
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(kStart)
        _bridge.deregisterHandler(kStop)
        _bridge.deregisterHandler(kIncrementMetric)
        _bridge.deregisterHandler(kPutMetric)
        _bridge.deregisterHandler(kGetLongMetric)
    }

    func start() {
        _trace.start()
    }

    func stop() {
        _trace.stop()
    }

    func incrementMetric(_ name: String, _ value: Int64) {
        _trace.incrementMetric(name, by: value)
    }

    func putMetric(_ name: String, _ value: Int64) {
        _trace.setValue(value, forMetric: name)
    }

    func getLongMetric(_ name: String) -> Int64 {
        return _trace.valueForMetric(name)
    }
}
